docs/adr/0003-version-strategy.md

```diff
@@ -0,0 +1,115 @@
+# Versioning for Nimbus ExperimentSchema
+
+- Status: Draft
+- Author: Kate Hudson
+- Deciders: Nimbus team, Cirrus team
+- Date: October 2020
+
+## Context and Problem Statement
+
+As per ADR-0002, we decided to use
+[JSON Schema Draft-07](https://tools.ietf.org/html/draft-handrews-json-schema-01) for the Nimbus
+Experiment Schema. Besides having some way of validating data across these applications, we also
+have to manage the additional complexity that any given time experiments may be live that are only
+compatible with a subset of clients.
+
+In order to prevent clients from receiving experiment defitions that could cause breakage while
+preserving the flexibility to update the Nimbus Experiment Schema when necessary, we need:
```

Comment:
This feeds into a broader point that I raised with @k88hudson in slack, but would welcome further opinions on: I think it's hard to define a meaningful versioning policy without understanding what you're trying to communicate or prevent; IOW you need to define what "breaking" means before you can determine whether something is a "breaking change".

For SemVer, "breaking" means "if a consumer updates to this version, there is some chance that their code will no longer compile/run correctly". What does "breaking" mean for the consumers of these experiment definitions?

For the Client SDK I can think of several:

1. A client that we want to have enrolled in the experiment, fails to do so, potentially skewing the data.
2. A client that we didn't want to enroll in the experiment, does so, potentially skewing the data.
3. A client trying to interpret this experiment info will crash or otherwise behave badly.

I don't think versioning is a good mechanism to prevent (3), which seems better served by having clients aggressively ignore experiment records that don't validate against their expected schema. But there does seem to be value in using a version number to manage (1) and (2).

I'm skeptical of needing three levels of versioning, at least from the client SDK perspective. A `MAJOR.MINOR` schema (or perhaps `MODEL.REVISION` in the style of schemaver) seems sufficient. Something like:

* Is the `MAJOR` number on this record larger than the one we support? Ignore it, we might do the wrong thing even if we manage to parse it correctly.
* Is the `MINOR` number on this record larger than the one we support? Attempt to validate/parse/execute it, but log and ignore if that fails.
* Otherwise, attempt to validate/parse/execute it and treat failure to do so as a noteworthy error for telemetry etc.

The third level of versioning may be useful for other consumers though, in which case I've no objection to it.